fadecandy_driver/src/fadecandy_driver/fadecandy_node.py

```diff
@@ -7,35 +7,55 @@
 
 class FadecandyNode:
     def __init__(self):
+        self._driver = None
+        self._diagnostic_updater = diagnostic_updater.Updater()
+        self._diagnostic_updater.add("Info", self._info_diagnostics)
+
+        self._connect()
+        self._set_leds_sub = rospy.Subscriber('set_leds', LEDArray, self._set_leds)
+        self._diagnostic_timer = rospy.Timer(rospy.Duration(1), lambda e: self._diagnostic_updater.force_update())
+
+    def _connect(self):
+        if self._driver is not None:
+            self._driver.release()
+            self._driver = None
+
+        rospy.loginfo('Connecting to Fadecandy device ..')
+
+        first_try = True
         connection_retry_rate = rospy.Rate(1.0)
         while not rospy.is_shutdown():
             try:
                 self._driver = FadecandyDriver()
             except IOError as e:
-                rospy.logwarn_once('Failed to connect to Fadecandy device: %s; will retry every second', e)
+                if first_try:
+                    rospy.logwarn('Failed to connect to Fadecandy device: %s; will retry every second', e)
```

Comment:
In my case, the diagnostics did not update when we came here. I cannot see an entry in `runtime_monitor`.
I believe that's because we are running `self._connect()` on line 14 before the diagnostics timer. Is that intentional @reinzor?

On a side note, it never reconnects in my case. I would have to restart the node to get that fixed. This is an ongoing problem.

Reply:
The new approach is also similar to `rospy.logwarn_once`. Is there a reason why we are switching to an if statement?

Reply:
> In my case, the diagnostics did not update when we came here. I cannot see an entry in `runtime_monitor`.
> I believe that's because we are running `self._connect()` on line 14 before the diagnostics timer. Is that intentional @reinzor?

Maybe it is better indeed to call the `_connect` after the diagnostics and before the `set_leds` publisher. This way you indeed have diagnostics before the initial connect

> On a side note, it never reconnects in my case. I would have to restart the node to get that fixed. This is an ongoing problem.

Strange, I've tested this. Will retest this tomorrow

> The new approach is also similar to rospy.logwarn_once. Is there a reason why we are switching to an if statement?

This way, you always show the `warn` message on a reconnection. Otherwise, it is only shown on the first connection. Provides a little bit more info to the user.